File: Ruby/Civitas/lib/sorpresa_ir_a_casilla.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

module Civitas
  class SorpresaIrACasilla < Sorpresa
    def initialize(tablero, valor, texto)
      super()
      @valor = valor
      @texto = texto
      @tablero = tablero
    end
  
    # Método sobrescrito
    def aplicar_a_jugador(actual, todos)
      if(jugador_correcto(actual, todos))  
        casilla=todos[actual].get_num_casilla_actual
        tirada=@tablero.calcular_tirada(casilla, @valor)
        nueva_pos=@tablero.nueva_posicion(casilla, tirada)
        todos[actual].mover_a_casilla(casilla)
        nueva_casilla=@tablero.get_casilla(nueva_pos)
        nueva_casilla.recibe_jugador(actual, todos)
        informe(actual, todos)
      end
    end
  
    def to_s
     "\n\t ¡SORPRESA! \n
       \n Has recibido: 
       \n #{@texto} \n"
    end
  end
end

File: Java/P5/Civitas/src/civitas/SorpresaIrACasilla.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

import java.util.ArrayList;

/**
 *
 * @author arturo
 */
public class SorpresaIrACasilla extends Sorpresa{
    private Tablero tablero;
    private int numCasilla;
    
    SorpresaIrACasilla(Tablero tablero, int numCasilla, String texto){
        super(texto);
        this.tablero=tablero;
        this.numCasilla=numCasilla;
    }
    
    @Override 
    public void aplicarAJugador(int actual, ArrayList<Jugador> todos){
        if(this.jugadorCorrecto(actual, todos)){
            informe(actual, todos);
            
            int casilla_actual=todos.get(actual).getNumCasillaActual();
            int casilla=(int)(Math.random()*20+1);
            int nueva_pos=tablero.nuevaPosicion(casilla_actual, casilla);
            
            todos.get(actual).moverACasilla(nueva_pos);
            tablero.getCasilla(nueva_pos).recibeJugador(actual, todos);
        }
    }
    
    @Override
    public String toString(){
        return "Sorpresa: "+super.getTexto()+"\nVas a la casilla: "+numCasilla+"\n";
    }
}
